lib/ffi.cpp

```diff
@@ -220,34 +233,95 @@ Value FFI_Pointer_read_string(Env *env, Value self, Args args, Block *) {
         auto length = args.at(0)->as_integer_or_raise(env)->to_nat_int_t();
         if (length < 0 || (size_t)length > std::numeric_limits<size_t>::max())
             env->raise("ArgumentError", "length out of range");
-        return new StringObject { (char *)address, (size_t)length };
+        return new StringObject { (char *)address, (size_t)length, Encoding::ASCII_8BIT };
     }
 
-    return new StringObject { (char *)address };
+    return new StringObject { (char *)address, Encoding::ASCII_8BIT };
 }
 
 Value FFI_Pointer_to_obj(Env *env, Value self, Args args, Block *) {
     args.ensure_argc_is(env, 0);
     return (Object *)self.send(env, "address"_s)->as_integer_or_raise(env)->to_nat_int_t();
 }
 
-Value FFI_MemoryPointer_initialize(Env *env, Value self, Args args, Block *) {
-    args.ensure_argc_is(env, 1);
+Value FFI_Pointer_write_string(Env *env, Value self, Args args, Block *) {
+    args.ensure_argc_between(env, 1, 2);
+
+    auto str = args.at(0)->as_string_or_raise(env);
+    auto length = args.size() > 1 ? IntegerObject::convert_to_native_type<size_t>(env, args.at(0)) : str->bytesize();
+
+    auto address = (void *)self.send(env, "address"_s)->as_integer_or_raise(env)->to_nat_int_t();
+    memcpy(address, str->c_str(), length);
 
-    auto size = args.at(0);
-    size->assert_type(env, Object::Type::Integer, "Integer");
-    self->ivar_set(env, "@size"_s, size);
+    return self;
+}
 
-    auto address = malloc(size->as_integer()->to_nat_int_t());
+Value FFI_MemoryPointer_initialize(Env *env, Value self, Args args, Block *block) {
+    args.ensure_argc_between(env, 1, 3);
+
+    size_t size = 0;
+    switch (args.at(0)->type()) {
+    case Object::Type::Symbol: {
+        auto sym = args.at(0)->as_symbol();
+        if (sym == "char"_s || sym == "uchar"_s || sym == "int8"_s || sym == "uint8"_s) {
+            size = 1;
+        } else if (sym == "short"_s || sym == "ushort"_s || sym == "int16"_s || sym == "uint16"_s) {
+            size = 2;
+        } else if (sym == "int32"_s || sym == "uint32"_s || sym == "float"_s) {
+            size = 4;
+        } else if (sym == "int64"_s || sym == "uint64"_s || sym == "long_long"_s || sym == "ulong_long"_s || sym == "double"_s) {
+            size = 8;
+        } else if (sym == "int"_s) {
+            size = sizeof(signed int);
+        } else if (sym == "uint"_s) {
+            size = sizeof(unsigned int);
```

Comment:
I'm not completely sure, but I would think the standards define that the signed and unsigned pairs have the same size, so these could be combined:
```c++
} else if (sym == "int"_s || sym == "uint"_s) {
    size = sizeof(int);
```

Reply:
I don't mind the duplication in this case.